### ISSUE TITLE
GroupEffects and onfinish handler of individual keyframe effects

### ISSUE THREAD

**Issue by hetsch** [web-animations/web-animations-js#110]
This might be more of a feature question but i could not find any hints in the documentation about this issue. Let say I have a `GroupEffect` holding some `KeyframeEffect` instances and passing the group to a `Animation` instance.

    let keyframeEffects = [
        new KeyframeEffect( ... ),
        new KeyframeEffect( ... ),
        ...
    ];

    let groupEffect = new GroupEffect(keyframeEffects);

    let animation = new Animation(groupEffect, document.timeline);
    animation.onfinish = this.onAnimationFinished;  // fired when all keyframe effects have finished
    animation.pause();

This works fine right now and my finish handler get's invoked if the whole animation is finished. How may one add a finish handler to the particular keyframe effects? The `KeyframeEffect` interface seems to have no `onfinished` handler. Is this by design or am I am overlooking something terribly. Thank's a lot!

**Comment by birtles** [comment]
Yes, that's by design since a keyframe effect in a group effect may finish many times. For example, consider a `GroupEffect` that contains a `KeyframeEffect` where the group effect has a spring timing function applied to it. As the spring comes to rest the keyframe effect would cross its "finished" boundary backwards and forwards many times. Requiring implementations to calculate all those crossings is very difficult to do in an efficient manner that preserves constant-time seeking.

**Comment by hetsch** [comment]
@birtles Thank you for your quick response, that make total sense for continues or springing animations. But think about - for example - a dropdown menu whose code might look like this:

    let effect = new SequenceEffect([
        // First animate the background
        new KeyframeEffect({
            dropdownBackground, [
                { transform: `scaleY(0)`, opacity: 0 },
                { transform: `scaleY(1)`, opacity: 1 }
            ], {
                duration: 1000,
                iterations: 1,
                fill: 'forwards'
            }   
        }),
        // Afterwords slide the dropdown items in (might be constructed in a loop)
        new GroupEffect({
            new KeyframeEffect({
                item1, [
                    { transform: 'translateY(-20px)', opacity: 0 },
                    { transform: 'translateY(0)', opacity: 1 }
                ], {
                    duration: 1000,
                    iterations: 1,
                    fill: 'forwards'
                    delay: 10
                }   
            }),
            new KeyframeEffect({
                item2, [
                    { transform: 'translateY(-20px)', opacity: 0 },
                    { transform: 'translateY(0)', opacity: 1 }
                ], {
                    duration: 1000,
                    iterations: 1,
                    fill: 'forwards'
                    delay: 20
                }   
            }),

            [...]

        })
    ])

    let animation = new Animation(effect, document.timeline);

The good think about this construct is that if the dropdown menu should expand one can call `animation.start()`, to collapse the whole thing simply calling `animation.reverse()` should be enough. I know this is a much simpler animation than you think about but here one might want to have a finish handler on the KeyframeEffects of `item1` and `item2` to apply another style or ... Or am I'm on the wrong track here and can you perhaps think of a better, easier way to achieve this? Thank's again.

**Comment by anderflash** [comment]
One interesting use case of a possible `finish` event of a KeyframeEffect could be a slide deck animation. We would like to have a _prezi_ effect: by moving a `div` called _canvas _ to show different content which is spread through the canvas. But we would like to stop at each keyframe and continue animation through a key press. In this case, instead of a current SequenceEffect, I think we need to manually develop a set of functions to control a collection of Animations.
